Introduce contracts commands to Rover docs (#1473)
apollographql/apollo-cli: docs/source/commands/contracts.mdx

---
title: Rover contract commands
description: For use with Apollo Contract variants
---

A **contract** is a variant that is the result of filtering out schema elements from a source variant.

Rover commands that interact with [contracts](https://www.apollographql.com/docs/graphos/delivery/contracts/) begin with `rover contract`.

## Publishing a contract to Apollo Studio

### `contract publish`

> This command requires [authenticating Rover with Apollo Studio](../configuring/#authenticating-with-apollo-studio).

You can use Rover to publish a new contract or publish configuration changes to an existing contract.

Run the `contract publish` command, like so:

```bash
rover contract publish my-graph@my-contract-variant \
  --source-variant my-source-variant \

>   --include foo --include bar \

Reviewer: ```suggestion
  --include foo \
  --include bar \
```

I think it helps to break these out onto separate lines for readability

  --exclude baz \
  --hide-unreachable-types
```

The argument `my-graph@my-contract-variant` in the example above is a [graph ref](../conventions/#graph-refs) that specifies the ID of the Studio graph you're publishing to, along with which [variant](/graphos/graphs/overview/#variants) you're publishing to.

If this variant already exists in the graph registry, its configuration is updated; otherwise, a new contract variant is created.

Options include:

<table class="field-table">
<thead>
<tr>
<th>Name</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr class="required">
<td>

###### `--source-variant`

</td>
<td>

The name of the [source variant](https://www.apollographql.com/docs/graphos/delivery/contracts/#2-fed1-enable-variant-support-for-tag) whose supergraph schema is used for filtering.

> This variant must be on the same graph as the contract variant and cannot be changed after contract creation. It must also be a federated variant with subgraphs, and if using Fed 1, support for the `@tag` directive must be explicitly enabled in Studio.

Reviewer: > This variant must be on the same graph as the contract variant and cannot be changed after contract creation.

What can't be changed here? The source variant? I'm having a hard time parsing this out.

Author: Yes this should be the source variant. Once a contract variant is made, you cannot change it's source variant. I can reword this so it is easier to parse.

> This variant must be on the same graph as the contract variant and cannot be changed after contract creation. It must also be a federated variant with subgraphs, and if using Fed 1, support for the `@tag` directive must be explicitly enabled in Studio.

Reviewer: I think it makes sense to split this up. Maybe even move the federation 1 bit into a sub-note in small italics.

```suggestion
This variant must be on the same graph as the contract variant and cannot be changed after contract creation. It must also be a federated variant with subgraphs. If your graph is using Federation 1, support for the `@tag` directive must be explicitly enabled in Studio.
```

**Required** the first time you publish a contract.

> **Optional** after your first publish, but if provided it must match the source variant in Studio.

Reviewer: ```suggestion
**Optional** after your first publish, but if provided it must match the previously published source variant in Studio.
```

I think this makes it a bit more clear

---
title: Rover contract commands
description: For use with Apollo Contract variants
---

A **contract** is a variant that is the result of filtering on a source variant:

Rover commands that interact with contracts begin with `rover contract`.

> ## Describing a contract variant

Reviewer: What do you think about putting `contract publish` _before_ `contract describe`? My reasoning is that unless you already have a contract in place, `contract describe` won't work. For folks reading docs on this for the first time it would make more sense to `contract publish` and _then_ `contract describe` in order to see what was just published.

Author: yeah i like the idea of putting `publish` before `describe`! i'll reorder it :)

These commands enable you to describe a contract in a federated graph.

### `contract describe`

> TODO(caydie) is this correct? I'm assuming you always have to authn to rover with studio 

Reviewer: Yes, this is correct. You always have to authenticate Rover with Studio when a graph ref is provided. Commands with no graph ref do not require auth (`rover dev` and `rover supergraph compose` are the main ones but there are some others).

Author: thank you!!

> This command requires [authenticating Rover with Apollo Studio](../configuring/#authenticating-with-apollo-studio).

You can use Rover to describe the contract variant of any source variant that Rover has access to.

Run the `contract describe` command, like so:

```bash
rover contract describe my-graph@my-contract-variant
```

The argument `my-graph@my-contract-variant` in the example above is a [graph ref](../conventions/#graph-refs) that specifies the ID of the Studio graph you're fetching from, along with which [variant](/graphos/graphs/overview/#variants) you're fetching.

> The output will contain information of how the contract variant was configured such as the source variant, include and exclude tags.

Reviewer: I think that once we have an example of what this output looks like (when the command is implemented) we should include that here

Author: yes i'm pairing with sachin later today to get an example output! will include that in the next commit

### `contract publish`

> This command requires [authenticating Rover with Apollo Studio](../configuring/#authenticating-with-apollo-studio).

You can use Rover to publish configuration changes to a contract.

Use the `contract publish` command, like so:

> TODO(caydie) is it my graph or my supergraph and what would be a good example for the arguments?

Reviewer: `my-graph` is fine here. I'm not entirely sure on good examples for args for this actually! cc @sachindshinde @jsegaran on that one

```bash
rover contract publish my-graph@my-contract-variant \
  --include [] \
  --exclude [] \
  --hide-unreachable-types
```

TODO(caydie): is it my-graph or is it my supergraph? i think it's my-graph
The argument `my-graph@my-variant` in the example above is a [graph ref](../conventions/#graph-refs) that specifies the ID of the Studio graph you're publishing to, along with which [variant](/graphos/graphs/overview/#variants) you're publishing to.

> If the graph and source variant exists in the graph registry but the contract variant does _not_, a new contract variant is created on publish.

Reviewer: maybe a silly question but how do you specify the source variant in this command? if i run the following:

```
rover contract publish my-graph@my-contract-variant \
  --include [] \
  --exclude [] \
  --hide-unreachable-types
```

it looks like i'm only specifying the contract variant - how do we know what the source variant should be in this scenario?

Author: you do indeed need to specify the `--source-variant` as an argument. we updated the doc to reflect that in the next commit!

> TODO(caydie) is the rest at the bottom relevant?

Reviewer: it looks like this was maybe copy pasted from another command page. we will need the table filled out with the arguments to this command instead of that other command. so we'll need an entry for `--include`, `--exclude`, `--hide-unreachable-types` and any other arguments we end up adding (`--source-variant`?)
